### PR TITLE
[api][filebrowser] Refactor and improve file upload public API core design

### PR DESCRIPTION
**Pull request by Harshg999**
## Overview

This PR introduces a new unified file upload REST API that provides a consistent interface for uploading files across multiple cloud storage providers. The implementation replaces the legacy upload handlers with modern, streaming-based handlers that offer improved performance, better error handling, and enhanced security.

### 1. **New REST API Endpoint** 
- **Endpoint**: `/api/v1/storage/upload/file/`
- **Class**: `UploadFileAPI` using Django REST Framework
- **Method**: POST with multipart/form-data support
- **Parameters**:
  - `destination_path` (required): Target path for the upload
  - `overwrite` (optional, default: false): Whether to overwrite existing files

### 2. **Unified Upload Handler Architecture**
- Introduced `get_upload_handler()` method in all filesystem interfaces
- Dynamic handler selection based on storage type and destination path
- Consistent interface across all storage providers

### 3. **Storage Provider Support**
Enhanced upload handlers for all major storage providers:
- **Amazon S3** (`S3NewFileUploadHandler`): Multipart streaming upload
- **Azure Blob Storage** (`ABFSNewFileUploadHandler`): Direct streaming with append operations
- **Google Cloud Storage** (`GSNewFileUploadHandler`): Multipart upload support
- **Apache Ozone** (`OFSNewFileUploadHandler`): Temporary file buffering approach
- **HDFS**: Enhanced with improved chunking support

### Security Enhancements
- **File Extension Validation**: Centralized validation in `is_file_upload_allowed()` function
- **Path Traversal Protection**: Validates filenames don't contain path separators
- **Permission Checks**: Verifies write access before upload initiation
- **Size Limits**: Enforces `MAX_FILE_SIZE_UPLOAD_LIMIT` during streaming

### Performance Optimizations
- **Streaming Uploads**: Direct streaming to cloud storage without full buffering
- **Configurable Chunk Sizes**: Provider-specific optimal chunk sizes
- **Multipart Support**: Parallel part uploads for S3 and GCS
- **Memory Efficiency**: Minimal memory footprint for large file uploads

### Error Handling
- **HTTP Status Codes**: Proper status codes for different error scenarios:
  - 400: Invalid parameters or file extensions
  - 403: Insufficient permissions
  - 404: Destination path not found
  - 409: File exists (when overwrite=false)
  - 413: File size exceeds limit
  - 500: Server errors
- **Detailed Error Messages**: User-friendly error descriptions
- **Automatic Cleanup**: Removes partial uploads on failure

### Validation & Testing
- **Request Validation**: New `UploadFileSerializer` for request parameter validation
- **Comprehensive Test Suite**: 
  - Unit tests for serializer validation
  - API tests with mocked filesystems
  - Edge case handling (empty files, special characters, etc.)

### Upload Handler Comparison

| Storage | Handler | Upload Method | Temporary Storage |
|---------|---------|---------------|-------------------|
| S3 | S3NewFileUploadHandler | Multipart streaming | None |
| Azure | ABFSNewFileUploadHandler | Append streaming | None |
| GCS | GSNewFileUploadHandler | Multipart streaming | None |
| Ozone | OFSNewFileUploadHandler | Buffered upload | Local temp file |
| HDFS | Enhanced existing | Chunked streaming | None |

This PR significantly improves the file upload experience in Hue while maintaining backward compatibility through the legacy endpoints.

### CODE REVIEW

**Review by JohanAhlen — APPROVED** [review]
Nice refactoring!

File: apps/filebrowser/src/filebrowser/api.py [hunk 1]
```diff
@@ -511,67 +516,59 @@ def upload_complete(request):
...
 
-  response = {
-    'uploaded_file_stats': _massage_stats(request, stat_absolute_path(filepath, request.fs.stats(filepath))),
-  }
+    fs = get_user_filesystem(username)
 
-  return JsonResponse(response)
+    upload_handler = fs.get_upload_handler(destination_path, overwrite)
+    if not upload_handler:
+      raise status.HTTP_400_BAD_REQUEST(f"No supported upload handler found for path: {destination_path}")
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply apps/filebrowser/src/filebrowser/api.py]: Done

File: apps/filebrowser/src/filebrowser/utils.py [hunk 2]
```diff
@@ -22,15 +22,28 @@
...
 from filebrowser.conf import ALLOW_FILE_EXTENSIONS, ARCHIVE_UPLOAD_TEMPDIR, RESTRICT_FILE_EXTENSIONS
+from filebrowser.lib.rwx import filetype, rwx
 
 LOG = logging.getLogger()
 
 
 DEFAULT_WRITE_SIZE = 1024 * 1024 * 128
 
 
+def get_user_filesystem(username: str) -> ProxyFS:
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply apps/filebrowser/src/filebrowser/utils.py]: Done

File: desktop/libs/azure/src/azure/abfs/abfs.py [hunk 3]
```diff
@@ -676,11 +673,109 @@ def _local_copy_file(self, local_src, remote_dst, chunk_size=UPLOAD_CHUCK_SIZE):
...
+          # Recursively check parent write access
+          return self.check_access(parent, permission="WRITE")
+
+        # For existing paths
+        if self.isdir(path):
+          # For directories, try creating a temporary marker file
+          temp_file = None
+          try:
+            # Generate unique temporary filename with timestamp
+            temp_file = self.join(path, ".hue_access_check_%s_%s" % (str(int(time.time() * 1000)), str(uuid.uuid4())[:8]))
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/libs/azure/src/azure/abfs/abfs.py]: Done

File: desktop/libs/hadoop/src/hadoop/fs/upload.py [hunk 4]
```diff
@@ -417,132 +417,111 @@ def file_complete(self, file_size):
...
-  def __del__(self):
-    if self._do_cleanup:
-      # Do not do cleanup here. It's hopeless. The self._fs threadlocal states
-      # are going to be all wrong.
-      LOG.debug(f"Check for left-over upload file for cleanup if the upload op was unsuccessful: {self._path}")
+      raise PopupException(f"Failed to write upload data: {e}", error_code=500)
 
-  def get_temp_path(self):
-    return self._path
+  def file_complete(self, file_size):
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/libs/hadoop/src/hadoop/fs/upload.py]: raw_data is based on the chunk size we can set for the HDFS handler and that is configurable in Hue for HDFS.

File: desktop/core/src/desktop/lib/fs/ozone/upload.py [hunk 5]
```diff
@@ -281,46 +284,174 @@ def _get_scheme(self):
...
+      if self.overwrite:
+        self._fs.remove(self.target_file_path, skip_trash=True)
+      else:
+        raise PopupException("File already exists: %s" % self.target_file_path, error_code=409)
+
+    # Create a temporary file in the configured temp directory to buffer the upload data
+    try:
+      self._temp_file = tempfile.NamedTemporaryFile(
+        mode="wb",
+        dir=self._fs._temp_dir,
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/core/src/desktop/lib/fs/ozone/upload.py]: Created a class property

File: desktop/libs/azure/src/azure/abfs/upload.py [hunk 6]
```diff
@@ -256,49 +258,106 @@ def _get_scheme(self):
...
+      raise PopupException(f"File exceeds maximum allowed size of {max_size} bytes.", error_code=413)
+
+    # Upload the chunk
+    self.upload_chunk(raw_data, start)
+    return None
 
+  def upload_chunk(self, raw_chunk, start):
+    try:
+      buffered_data = BytesIO(raw_chunk)
+      self._fs._append(self.target_path, buffered_data, params={"position": int(start)})
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/libs/azure/src/azure/abfs/upload.py:382]: Not changing this in the old code until it is fully refactored. For now, leaving it as it is but adding a comment for future.

File: desktop/libs/aws/src/aws/s3/upload.py [hunk 7]
```diff
@@ -244,44 +246,97 @@ def _get_file_part(self, raw_data):
...
+      raise PopupException("Invalid filename. Path separators are not allowed.", error_code=400)
+
+    # Check if the user has write access to the destination path
+    if not self._fs.check_access(self.dest_path, permission='WRITE'):
+      raise PopupException(f"Insufficient permissions to write to S3 path {self.dest_path}.", error_code=403)
+
+    # Check if the file already exists at the destination path
+    if self._fs.exists(self._fs.join(self.dest_path, file_name)):
+      if self.overwrite:
+        self._fs.remove(self._fs.join(self.dest_path, file_name))
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/libs/aws/src/aws/s3/upload.py]: Done

File: desktop/libs/hadoop/src/hadoop/fs/upload.py [hunk 8]
```diff
@@ -417,132 +417,111 @@ def file_complete(self, file_size):
...
-      LOG.error("Not using HDFS upload handler: %s" % (ex))
-      raise ex
+    # Check file extension restrictions
+    is_allowed, err_message = is_file_upload_allowed(file_name)
+    if not is_allowed:
+      raise PopupException(err_message, error_code=400)
 
-    raise StopFutureHandlers()
+    # Check if the destination path already exists or not
+    if not self._fs.exists(self.dest_path):
```
[the remark this answers is not in the document]
Comment (Harshg999) [reply desktop/libs/hadoop/src/hadoop/fs/upload.py:493]: The checks are fine, ignoring this one.

File: apps/filebrowser/src/filebrowser/utils.py [hunk 9]
```diff
@@ -22,15 +22,28 @@
...
 from filebrowser.conf import ALLOW_FILE_EXTENSIONS, ARCHIVE_UPLOAD_TEMPDIR, RESTRICT_FILE_EXTENSIONS
+from filebrowser.lib.rwx import filetype, rwx
 
 LOG = logging.getLogger()
 
 
 DEFAULT_WRITE_SIZE = 1024 * 1024 * 128
 
 
+def get_user_fs(username: str) -> ProxyFS:
```
Comment (agl29) [inline apps/filebrowser/src/filebrowser/utils.py:37]: Can you please add a comment in the function explaining its purpose and why it is needed?
for eg if there are 2 fs then what is the user FS

File: apps/filebrowser/src/filebrowser/utils.py [hunk 10]
```diff
@@ -168,3 +181,15 @@ def is_file_upload_allowed(file_name):
       return False, f'File type "{file_type}" is restricted. Update file extension restrictions to allow this type.'
 
   return True, None
+
+
+def massage_stats(stats):
```
Comment (agl29) [inline apps/filebrowser/src/filebrowser/utils.py:205]: Same here, please add an explanation of what this function does.

File: desktop/core/src/desktop/api_public_urls_v1.py [hunk 11]
```diff
@@ -114,7 +115,7 @@
   re_path(r'^storage/rename/?$', api_public.storage_rename, name='storage_rename'),
   re_path(r'^storage/move/?$', api_public.storage_move, name='storage_move'),
   re_path(r'^storage/copy/?$', api_public.storage_copy, name='storage_copy'),
-  re_path(r'^storage/upload/file/?$', api_public.storage_upload_file, name='storage_upload_file'),
+  re_path(r'^storage/upload/file/?$', filebrowser_api.UploadFileAPI.as_view(), name='storage_upload_file'),
```
Comment (agl29) [inline desktop/core/src/desktop/api_public_urls_v1.py:118]: why we are not defining this function in api_public file, like earlier?

File: desktop/core/src/desktop/lib/fs/gc/gs.py [hunk 12]
```diff
@@ -479,12 +479,6 @@ def _check_key_parent_path(self, src, dst):
...
-  @auth_error_handler
-  def upload_v1(self, META, input_data, destination, username):
-    from desktop.lib.fs.gc.upload import GSNewFileUploadHandler  # Circular dependency
-
-    gs_upload_handler = GSNewFileUploadHandler(destination, username)
-
-    parser = MultiPartParser(META, input_data, [gs_upload_handler])
-    return parser.parse()
+  def get_upload_handler(self, destination_path, overwrite):
+    from desktop.lib.fs.gc.upload import GSNewFileUploadHandler
```
Comment (agl29) [inline desktop/core/src/desktop/lib/fs/gc/gs.py:479]: should we add this comment again? # Circular dependency

File: desktop/core/src/desktop/lib/fs/gc/upload.py [hunk 13]
```diff
@@ -186,48 +189,91 @@ def _get_file_part(self, raw_data):
...
+    # Check if the user has write access to the destination path
+    if not self._fs.check_access(self.dest_path, permission='WRITE'):
+      raise PopupException(f"Insufficient permissions to write to GS path {self.dest_path}.", error_code=403)
+
+    # Check if the file already exists at the destination path
+    if self._fs.exists(self._fs.join(self.dest_path, file_name)):
+      if self.overwrite:
+        self._fs.remove(self._fs.join(self.dest_path, file_name))
+      else:
+        raise PopupException(f"The file {file_name} already exists at the destination path.", error_code=409)
```
Comment (agl29) [inline desktop/core/src/desktop/lib/fs/gc/upload.py:297]: Can we move the code block from lines 214 to 240 into a separate function to keep the new_file function cleaner?

File: desktop/libs/aws/src/aws/s3/upload.py [hunk 14]
```diff
@@ -244,44 +246,98 @@ def _get_file_part(self, raw_data):
...
+    if not self._fs.check_access(self.dest_path, permission='WRITE'):
+      raise PopupException(f"Insufficient permissions to write to S3 path {self.dest_path}.", error_code=403)
+
+    # Check if the file already exists at the destination path
+    file_path = self._fs.join(self.dest_path, file_name)
+    if self._fs.exists(file_path):
+      if self.overwrite:
+        self._fs.remove(file_path)
+      else:
+        raise PopupException(f"The file {file_name} already exists at the destination path.", error_code=409)
```
Comment (agl29) [inline desktop/libs/aws/src/aws/s3/upload.py:354]: same, code lines 270-298 to a separate fn?

File: desktop/libs/azure/src/azure/abfs/abfs.py [hunk 15]
```diff
@@ -676,11 +673,109 @@ def _local_copy_file(self, local_src, remote_dst, chunk_size=UPLOAD_CHUCK_SIZE):
...
+          return True
+        except WebHdfsException as e:
+          if e.code in (401, 403):  # Unauthorized or Forbidden
+            LOG.debug(f'No read permission for path "{path}": {str(e)}')
+            return False
+          # Re-raise unexpected errors
+          raise
+
+      # Check WRITE permission
+      else:
```
Comment (agl29) [inline desktop/libs/azure/src/azure/abfs/abfs.py:722]: should we add hard check here? permission==WRITE?

File: desktop/libs/azure/src/azure/abfs/upload.py [hunk 16]
```diff
@@ -256,49 +258,107 @@ def _get_scheme(self):
...
-      self.target_path = self._fs.join(self.destination, file_name)
+    # Check if the file name contains a path separator
+    # This prevents directory traversal attacks
+    if os.path.sep in file_name:
+      raise PopupException("Invalid filename. Path separators are not allowed.", error_code=400)
+
+    # Check if the user has write access to the destination path
+    if not self._fs.check_access(self.dest_path, permission="WRITE"):
+      raise PopupException(f"Insufficient permissions to write to ABFS path {self.dest_path}.", error_code=403)
+
```
Comment (agl29) [inline desktop/libs/azure/src/azure/abfs/upload.py:351]: same here from code lines 277-298 to a new separate fn?

**Review by agl29 — COMMENTED** [review]
Nice work!
I've left a few comments,  please take a look.

File: apps/filebrowser/src/filebrowser/utils.py (hunk 9 above)
Comment (Harshg999) [reply apps/filebrowser/src/filebrowser/utils.py:37]: Done

File: apps/filebrowser/src/filebrowser/utils.py (hunk 10 above)
Comment (Harshg999) [reply apps/filebrowser/src/filebrowser/utils.py:205]: Done

File: desktop/core/src/desktop/api_public_urls_v1.py (hunk 11 above)
Comment (Harshg999) [reply desktop/core/src/desktop/api_public_urls_v1.py:118]: We are slowly cleaning up the api_public file so that the specific code remains in the specific working module scope for better separation of concerns. 

The URL endpoint mapping for public APIs is still present in the single file for the time being.

File: desktop/core/src/desktop/lib/fs/gc/gs.py (hunk 12 above)
Comment (Harshg999) [reply desktop/core/src/desktop/lib/fs/gc/gs.py:479]: I think its not required and doesn't provide any meaningful value but open for discussion!

File: desktop/libs/azure/src/azure/abfs/abfs.py (hunk 15 above)
Comment (Harshg999) [reply desktop/libs/azure/src/azure/abfs/abfs.py:722]: Line 697-699 already handles that case

File: desktop/core/src/desktop/lib/fs/gc/upload.py (hunk 13 above)
Comment (Harshg999) [reply desktop/core/src/desktop/lib/fs/gc/upload.py:297]: Nice suggestion, fixed it

File: desktop/libs/azure/src/azure/abfs/upload.py (hunk 16 above)
Comment (Harshg999) [reply desktop/libs/azure/src/azure/abfs/upload.py:351]: Nice suggestion, fixed it

File: desktop/libs/aws/src/aws/s3/upload.py (hunk 14 above)
Comment (Harshg999) [reply desktop/libs/aws/src/aws/s3/upload.py:354]: Nice suggestion, fixed it

**Review by ramprasadagarwal — APPROVED** [review]
Nice refactor
